File: 03.programmers/lv01/01.C++/dart_game.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int solution(string dartResult) {
    int answer = 0;
    vector<int> tmp; 
    for(int i=0; i<dartResult.length(); i++){
        char c = dartResult[i];
        if(c>='0' && c<='9'){
            if(c == '1' && dartResult[i+1] == '0'){
                tmp.push_back(10);
                ++i;
            }
            else tmp.push_back(c - '0');
        }
        else{
            int back = tmp.size()-1;
            if(c == 'S') continue;
            else if(c == 'D') tmp[back] = tmp[back] * tmp[back];
            else if(c == 'T') tmp[back] = tmp[back] * tmp[back] * tmp[back];
            else if(c == '*'){
                if(back == 0) tmp[back] *=2;
                else{
                    tmp[back]*=2;
                    tmp[back-1]*=2;
                }
            }
            else if(c == '#')tmp[back] *=-1;
        }
    }
    for(auto res : tmp) answer += res;
    return answer;
}
int main(){
    string dartResult = {"1S2D*3T"};
    cout<<solution(dartResult)<<endl;

    return 0;
}

File: 03.programmers/lv01/02.Python/dart_game.py
def solution(dartResult):
    answer = []
    for i, c in enumerate(dartResult):
        if c.isdigit(): 
            if c==0 and dartResult[i-1] == 1:
                answer.pop(-1)
                answer.append(10)
            else:
                answer.append(int(c))
        else:
            if c == 'S': continue
            elif c == 'D': answer[-1] = answer[-1]**2
            elif c == 'T': answer[-1] = answer[-1]**3
            elif c == '#': answer[-1]*=-1
            elif c == '*':
                answer[-1] *=2
                if len(answer) > 1 : answer[-2] *=2
                    
    return sum(answer)

input_dartRsult = "1S2D*3T"
print(solution(input_dartRsult))
